### PR TITLE
feature: new xDhcpServerExclusionRange Resource

### LINKED ISSUE #7: Feature Request - Exclusion Range
Hi Team, please can you kindly add the ability to add/remove DHCP exclusion ranges. 

Example below; At the moment only Method 1 is available.

Say for instance Network Devices/Servers sit on 192.168.1.1 to 192.168.1.20

Method 1: Scope 192.168.1.21 -> 192.168.1.254

Method 2: Scope 192.168.1.1 -> 192.168.1.254
                 Reservation 192.168.1.1 -> 192.168.1.20

Many organizations use Method 2 as best practice for defining DHCP scopes.

DHCP cmdlets to accomplish this;

Get-DhcpServerv4ExclusionRange
Add-DhcpServerv4ExclusionRange
Remove-DhcpServerv4ExclusionRange

### PR DESCRIPTION
**Pull request by cdituri**
Should Close #7. Passes some basic manual testing, ie Ensure Absent/Present with single and multiple declarations, but would appreciate a review and second set of eyes (or better yet, a unit test if there's anyone out there who's braver / more willing to write one :-).

### CODE REVIEW

File: DSCResources/MSFT_xDhcpServerExclusionRange/MSFT_xDhcpServerExclusionRange.psm1 [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+    {
+        $errorMsg = $LocalizedData.InvalidStartAndEndRangeMessage
+        New-TerminatingError -errorId RangeNotCorrect -errorMessage $errorMsg -errorCategory InvalidArgument
+    }
+
+#endregion Input Validation
+
+    if ($AddressFamily -eq "IPv4")
+    {
+      $dhcpExclusionRange = Get-DhcpServerv4ExclusionRange | ? {($_.ScopeId -eq $ScopeId) -and ($_.StartRange -eq $StartRange) -and ($_.EndRange -eq $EndRange)}
```
Comment (iainbrighton) [inline DSCResources/MSFT_xDhcpServerExclusionRange/MSFT_xDhcpServerExclusionRange.psm1]: Can you replace the alias with Where-Object and place on a new line to try an keep the line length below 100 chars as per the [contribution guidelines](https://github.com/PowerShell/DscResources/blob/master/StyleGuidelines.md)?

File: DSCResources/MSFT_xDhcpServerExclusionRange/MSFT_xDhcpServerExclusionRange.psm1 [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+
+        [ValidateSet('IPv4')]
+        [String]$AddressFamily = 'IPv4',
+
+        [ValidateSet('Present','Absent')]
+        [String]$Ensure = 'Present'
+    )
+
+
+    if($PSBoundParameters.ContainsKey('Debug')){ $null = $PSBoundParameters.Remove('Debug')}
```
Comment (iainbrighton) [inline DSCResources/MSFT_xDhcpServerExclusionRange/MSFT_xDhcpServerExclusionRange.psm1]: Braces on a new line..

File: DSCResources/MSFT_xDhcpServerExclusionRange/MSFT_xDhcpServerExclusionRange.psm1 [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+    # Check to ensure startRange is smaller than endRange
+    if($EndRange.Address -lt $StartRange.Address)
+    {
+        $errorMsg = $LocalizedData.InvalidStartAndEndRangeMessage
+        New-TerminatingError -errorId RangeNotCorrect -errorMessage $errorMsg -errorCategory InvalidArgument
+    }
+
+#endregion Input Validation
+
+    if($PSBoundParameters.ContainsKey('Debug')){ $null = $PSBoundParameters.Remove('Debug')}
```
Comment (iainbrighton) [inline DSCResources/MSFT_xDhcpServerExclusionRange/MSFT_xDhcpServerExclusionRange.psm1]: Braces on a new line..

File: DSCResources/MSFT_xDhcpServerExclusionRange/MSFT_xDhcpServerExclusionRange.psm1 [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+        # If dhcpExclusionRange should be absent
+        else
+        {
+            if($Apply)
+            {
+                $removingExclusionRangeMsg = $LocalizedData.RemovingExclusionRangeMessage
+                Write-Verbose -Message $removingExclusionRangeMsg
+
+                # Remove the scope
+                Remove-DhcpServerv4ExclusionRange -ScopeId $dhcpExclusionRange.ScopeId
```
Comment (iainbrighton) [inline DSCResources/MSFT_xDhcpServerExclusionRange/MSFT_xDhcpServerExclusionRange.psm1]: Does Remove-DhcpServerv4ExclusionRange prompt for confirmation (it should do!)? If so, we'll need a -confirm:$false?

File: DSCResources/MSFT_xDhcpServerExclusionRange/MSFT_xDhcpServerExclusionRange.psm1 [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+        [ValidateSet('Present','Absent')]
+        [String]$Ensure = 'Present',
+
+        [Switch]$Apply
+    )
+
+    $checkExclusionRangeMessage = $LocalizedData.CheckExclusionRangeMessage
+    Write-Verbose -Message $checkExclusionRangeMessage
+
+    $dhcpExclusionRange = Get-DhcpServerv4ExclusionRange | ? {($_.ScopeId -eq $ScopeId) -and ($_.StartRange -eq $StartRange) -and ($_.EndRange -eq $EndRange)}
```
Comment (iainbrighton) [inline DSCResources/MSFT_xDhcpServerExclusionRange/MSFT_xDhcpServerExclusionRange.psm1]: Replace the alias and break onto a new line.

File: DSCResources/MSFT_xDhcpServerExclusionRange/MSFT_xDhcpServerExclusionRange.psm1 [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+    # dhcpExclusionRange is set
+    if($dhcpExclusionRange)
+    {
+        $TestExclusionRangeMessage = $($LocalizedData.TestExclusionRangeMessage) -f 'present', $Ensure
+        Write-Verbose -Message $TestExclusionRangeMessage
+
+        # if it should be present, test individual properties to match parameter values
+        if($Ensure -eq 'Present')
+        {
+            #region Test the ScopeId
```
Comment (iainbrighton) [inline DSCResources/MSFT_xDhcpServerExclusionRange/MSFT_xDhcpServerExclusionRange.psm1]: I'm not sure I follow the logic - would we ever get here? $dhcpExclusionRange will only ever _not be $null_ if the -ScopeId, StartRange _and_ EndRange are not defined. This is the sort of stuff unit tests would expose/test..

**Comment by iainbrighton** [comment]
@cdituri Looking pretty good :+1: . There's a query around the logic in Validate- function. Any chance of some unit tests to double-check the logic? :wink:

**Comment by cdituri** [comment]
@iainbrighton thanks for the review and fast response! I can fix the issues you mentioned in the inline comments, no prob. Any unit testing will probably have to wait til after the weekend, gotta keep the Mrs. happy first ;) Cheers.

**Comment by TravisEz13** [comment]
Thanks for your contribution but we need you to provide adequate test coverage for the code you change. [Contribution Guidelines -Adding test coverage for DSC resources](https://github.com/PowerShell/DscResources/blob/master/CONTRIBUTING.md#adding-test-coverage-for-dsc-resources)  Please work with us to come up with a plan to cover this code.

**Comment by cdituri** [comment]
@TravisEz13 - I can create code coverage for this new resource but its going to take some time (ie it must happen off-hours, on my own time). If you have any suggestions as to the minimum acceptable coverage required I'd be happy to have a go at it.

**Comment by iainbrighton** [comment]
@cdituri Start small, create a few tests targetting the Get- and Set- methods to begin with and we'll help you out. There are plenty of examples in the other resources/repos that you can look at too.

**Comment by cdituri** [comment]
@iainbrighton - borrowed the outline of `MSFT_xDhcpServerExclusionRange.test.ps1` from the existing Pester tests, just need to populate it now.

**Comment by TravisEz13** [comment]
@cdituri Good improvement.  You have the basic outline for the tests, but we need to test at least test-targetresource  and set-targetresource.

## Suggested Tests
1. for test-targetresource, the minimum would be to call the method, mocking the actual resource get, verify that  the mock is called and verify the correct result.  

2. For set-targetresource, the minimum would be similar, call the method, mock the actual resource calls,  and verify the mocks are called.

*Note, I'm open to other tests* 
As for as what parameters to use in the above tests, Writing tests for every possible parameter combination that causes a behavior change, is ideal, but start with the most common case.

This is very minimal coverage. 
## Getting help
Message me or other contributors to this if you get stuck.  Or ask questions on [Gitter dsc resources channel](https://gitter.im/PowerShell/DscResources)  The [PowerShell Slack channels](http://slack.poshcode.org/) are also a good place to get help.

## Summary 
Having basic tests will help verify that the resource basically behaves correctly, if someone needs to fix something in the resource and encourage others to contribute tests.

**Comment by TravisEz13** [comment]
If there is no update of the pull request within 9 days, we will close the pull request.  After that, if you have an update, feel free to reopen the pull request or a new pull request.  Please consider closing and re-opening when ready for review.

**Comment by cdituri** [comment]
@TravisEz13 - that makes sense. Closing for now but will reopen once the code has adequate test coverage.

**Comment by Zuldan** [comment]
Really looking forward to this resource. Keep up the good work cdituri. I need this one in production asap.

**Comment by TravisEz13** [comment]
@cdituri If there is no issue for this you should open an issue to let people know you are working on it.  Include a link to your fork and branch if you want help.  The best way we have for now to say we are waiting for you is to close the PR.

**Pull request closed without merging**